### ISSUE TITLE
Panic at startup when trying to get Datadog credentials

### ISSUE THREAD

**Issue by leosunmo** [DataDog/datadog-operator#662]
**Output of the info page (if this is a bug)**
Unsure what the info page is, but versions are available in the logs.

**Describe what happened:**
Panic during startup when trying to get credentials.
The datadog credentials are available to the operator as Environment variables through secrets, and they _seem_ to be mounted correctly.
```
<snip>
    Environment:
      WATCH_NAMESPACE:  datadog (v1:metadata.namespace)
      POD_NAME:         datadog-operator-5559bd6ddb-2657n (v1:metadata.name)
      DD_API_KEY:       <set to the key 'api-key' in secret 'datadog-secrets'>  Optional: false
      DD_APP_KEY:       <set to the key 'app-key' in secret 'datadog-secrets'>  Optional: false
<snip>
```

**Steps to reproduce the issue:**
Running the following Helm chart setup:
`datadog-operator` version `0.9.1`
Deployed via Flux HelmRelease (but not customized settings).
Values:
```
installCRDs: true
apiKeyExistingSecret: datadog-secrets
appKeyExistingSecret: datadog-secrets
```

**Additional environment details (Operating System, Cloud provider, etc):**
Running GKE _AutoPilot_ Cluster (AutoPilot does some weird things sometimes, but not sure it is to blame this time):
```
Client Version: v1.24.8
Kustomize Version: v4.5.4
Server Version: v1.24.8-gke.401
```

Here's the logs from the Datadog-operator at startup (spacing added by me to signify time diff. It's acquiring the lock it seems):

```
{"level":"INFO","ts":"2022-12-13T14:20:31Z","logger":"setup","msg":"Version: v0.8.1"}
{"level":"INFO","ts":"2022-12-13T14:20:31Z","logger":"setup","msg":"Build time: 2022-07-06/16:25:24"}
{"level":"INFO","ts":"2022-12-13T14:20:31Z","logger":"setup","msg":"Git Commit: 3bbf4a75bf80358815d76deef4f1cb3581ab43fe"}
{"level":"INFO","ts":"2022-12-13T14:20:31Z","logger":"setup","msg":"Go Version: go1.17.11"}
{"level":"INFO","ts":"2022-12-13T14:20:31Z","logger":"setup","msg":"Go OS/Arch: linux/amd64"}
{"level":"INFO","ts":"2022-12-13T14:20:31Z","logger":"setup","msg":"Manager will be watching namespace","namespace":"datadog"}
{"level":"INFO","ts":"2022-12-13T14:20:32Z","logger":"klog","msg":"Waited for 1.017309257s due to client-side throttling, not priority and fairness, request: GET:https://10.161.64.1:443/apis/source.toolkit.fluxcd.io/v1beta1?timeout=32s\n"}
{"level":"INFO","ts":"2022-12-13T14:20:32Z","logger":"controller-runtime.metrics","msg":"metrics server is starting to listen","addr":":8383"}
{"level":"INFO","ts":"2022-12-13T14:20:32Z","logger":"setup","msg":"Feature disabled, not starting the controller","controller":"DatadogMonitor"}
{"level":"INFO","ts":"2022-12-13T14:20:32Z","logger":"setup","msg":"starting manager"}
{"level":"INFO","ts":"2022-12-13T14:20:32Z","logger":"klog","msg":"attempting to acquire leader lease datadog/datadog-operator-lock...\n"}
{"level":"INFO","ts":"2022-12-13T14:20:32Z","logger":"controller-runtime.manager","msg":"starting metrics server","path":"/metrics"}



{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"klog","msg":"successfully acquired lease datadog/datadog-operator-lock\n"}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting EventSource","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","source":"kind source: /, Kind="}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Starting Controller","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent"}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"KubeAPIWarningLogger","msg":"policy/v1beta1 PodDisruptionBudget is deprecated in v1.21+, unavailable in v1.25+; use policy/v1 PodDisruptionBudget"}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"DatadogMetricForwarders","msg":"New Datadog metrics forwarder registred","ID":"datadog/datadog"}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"DatadogMetricForwarders","msg":"Starting Datadog metrics forwarder","CustomResource.Namespace":"datadog","CustomResource.Name":"datadog"}
{"level":"INFO","ts":"2022-12-13T14:21:53Z","logger":"DatadogMetricForwarders","msg":"Getting Datadog credentials","CustomResource.Namespace":"datadog","CustomResource.Name":"datadog"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1385de4]
goroutine 452 [running]:
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).getCredsFromDatadogAgent(0x0, 0x0)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:517 +0x24
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).getCredentials(0xc000850c60, 0x1b29dc4)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:499 +0x25
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).connectToDatadogAPI(0xc000850c60)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:209 +0xa5
k8s.io/apimachinery/pkg/util/wait.PollImmediateUntil(0xc000850c60, 0xc000765f30, 0xc00005f000)
	/go/pkg/mod/k8s.io/apimachinery@v0.21.2/pkg/util/wait/wait.go:500 +0x2f
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).start(0xc000850c60, 0xc000586fb8)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:136 +0xdf
created by github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*ForwardersManager).Register
	/workspace/pkg/controller/utils/datadog/forwarders_manager.go:65 +0x292
```

**Comment by leosunmo** [comment]
It looks like my `datadogagents.datadoghq.com` resource was being deleted perhaps? Hard to say, but I nuked it all and re-applied it and it is no longer panicking.

**Comment by CharlyF** [comment]
hi @leosunmo - Thanks for opening this up.
Could you share the DatadogAgent manifest? It's best to configure creds in the `spec.credentials` [section](https://github.com/DataDog/datadog-operator/blob/main/apis/datadoghq/v1alpha1/datadogagent_types.go#L38), it'll be easier to troubleshoot.
On a side note, I have to check but I think we don't support autopilot with the operator (at least in the 1.0). It's in our backlog, and I think it would fail later (when mounting volumes).

**Comment by leosunmo** [comment]
Yeah, the AutoPilot is very unhappy with the volume mounts, which is why I mentioned it :P 
```
datadog-operator {"level":"ERROR","ts":"2022-12-13T15:14:27Z","logger":"controller-runtime.manager.controller.datadogagent","msg":"Reconciler error","reconciler group":"datadoghq.com","reconciler kind":"DatadogAgent","name":"datadog","namespace":"datadog","error":"admission webhook \"gkepolicy.common-webhooks.networking.gke.io\" denied the request: GKE Warden rejected the request because it violates one or more constraints.\nViolations details: {\"[denied by autogke-no-write-mode-hostpath]\":[\"hostPath volume procdir used in container agent uses path /proc which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume cgroups used in container agent uses path /sys/fs/cgroup which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume runtimesocketdir used in container agent uses path /var/run which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume cgroups used in container process-agent uses path /sys/fs/cgroup which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume procdir used in container process-agent uses path /proc which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume runtimesocketdir used in container process-agent uses path /var/run which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume procdir used in container init-config uses path /proc which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume cgroups used in container init-config uses path /sys/fs/cgroup which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\",\"hostPath volume runtimesocketdir used in container init-config uses path /var/run which is not allowed in Autopilot. Allowed path prefixes for hostPath volumes are: [/var/log/].\"]}\nRequested by user: 'system:serviceaccount:datadog:datadog-operator', groups: 'system:serviceaccounts,system:serviceaccounts:datadog,system:authenticated'."}
```


This is the agent config I'm using.
```
apiVersion: datadoghq.com/v1alpha1
kind: DatadogAgent
metadata:
  name: datadog
  namespace: datadog
spec:
  site: datadoghq.eu
  # Credentials to communicate between:
  #  * Agents and Datadog (API/APP key)
  #  * Node Agent and Cluster Agent (Token)
  credentials:
    apiSecret:
      secretName: "datadog-secrets"
      keyName: "api-key"
    appSecret:
      secretName: "datadog-secrets"
      keyName: "app-key"
    
  # Node Agent configuration    
  agent:
    config:
      tolerations:
        - operator: Exists

  # Cluster Agent configuration
  clusterAgent:
    config:
      clusterChecksEnabled: true
    replicas: 2
```

**Comment by CharlyF** [comment]
okay so from the config standpoint, this is valid and I am not sure why it crashed. but that comment:
> It looks like my datadogagents.datadoghq.com resource was being deleted perhaps?
leads me to believe that maybe there was something else that played here. Do you want to keep this open in case you have more details? Otherwise I am happy to close in favor of the issue #663 that is the main blocker for your use case.

**Comment by jcogilvie** [comment]
I can confirm this is an issue for me when getting creds on 0.8.3.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1322664]

goroutine 644 [running]:
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).getCredsFromDatadogAgent(0x0?, 0x0?)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:517 +0x24
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).getCredentials(0xc00032f080, 0x1adc1a1?)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:499 +0x25
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).connectToDatadogAPI(0xc00032f080)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:209 +0xa5
k8s.io/apimachinery/pkg/util/wait.PollImmediateUntil(0xc00032f080?, 0xc00093ff30, 0x0?)
	/go/pkg/mod/k8s.io/apimachinery@v0.21.2/pkg/util/wait/wait.go:500 +0x2f
github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*metricsForwarder).start(0xc00032f080, 0x0?)
	/workspace/pkg/controller/utils/datadog/metrics_forwarder.go:136 +0xdf
created by github.com/DataDog/datadog-operator/pkg/controller/utils/datadog.(*ForwardersManager).Register
	/workspace/pkg/controller/utils/datadog/forwarders_manager.go:65 +0x28d
{"level":"INFO","ts":"2022-12-19T19:39:34Z","logger":"setup","msg":"Version: v0.8.3"}
{"level":"INFO","ts":"2022-12-19T19:39:34Z","logger":"setup","msg":"Build time: 2022-11-03/03:16:58"}
{"level":"INFO","ts":"2022-12-19T19:39:34Z","logger":"setup","msg":"Git Commit: 5004d77a95778ba6b1e4703208e8b996f5ee9bcd"}
{"level":"INFO","ts":"2022-12-19T19:39:34Z","logger":"setup","msg":"Go Version: go1.18.8"}
{"level":"INFO","ts":"2022-12-19T19:39:34Z","logger":"setup","msg":"Go OS/Arch: linux/amd64"}
{"level":"INFO","ts":"2022-12-19T19:39:34Z","logger":"setup","msg":"Manager will be watching namespace","namespace":"datadog"}
```

Secret in namespace `datadog` with name `datadog-secret` does exist and should be readable by the agent (and was working previously).

**Comment by jcogilvie** [comment]
I think this is happening for me at least when trying to apply a v2 DatadogAgent with `storage` still set to v1.

**Comment by bewinsnw** [comment]
For others hitting this: the code to read the v2 config doesn't exist in v0.8.3 or v0.8.4 (the current image tag used by default by the operator chart). The panic above is here:
https://github.com/DataDog/datadog-operator/blame/b3d4ea45f64bf4a8a8c71d6ecfe5c353d2dd9c3f/pkg/controller/utils/datadog/metrics_forwarder.go#L517

which is prior to this commit adding v2 support https://github.com/DataDog/datadog-operator/commit/f3377487609759797b62137a1b2da187f1639804

so you got here using the datadog-operator helm chart, you need to update image.tag to use a version >0.8.4 https://hub.docker.com/r/datadog/operator/tags

The actual error comes from the agent config moving the credentials to under the 'global.' prefix for v2.

**Comment by bewinsnw** [comment]
@CharlyF "okay so from the config standpoint, this is valid and I am not sure why it crashed." see also #711 . The crd shipped for datadogagents sets storage=true for v1alpha1 but _served_ true for v2alpha1, and since v2alpha1 > v1alpha1, kubectl (and presumably other bits of the k8s ecosystem) default to using the higher api version. So we get manifests back that say they are v2alpha1 but are in v1alpha1 format and are invalid. I could see that as the cause here, it's causing us a lot of problems (we're trying to jump up to v2 to avoid the issues this created with v1, and that's how I hit this bug)

**Comment by CharlyF** [comment]
Hi @bewinsnw - apologies for the long delay, we just released [the Datadog Operator 1.0](https://github.com/DataDog/datadog-operator/releases/tag/v1.0.1) yesterday and we also bumped the [helm chart to 1.0.0 ](https://github.com/DataDog/helm-charts/blob/main/charts/datadog-operator/CHANGELOG.md) which defaults the stored version to v2alpha1 and the operator to 1.0.
There is a [Note](https://github.com/DataDog/helm-charts/blob/main/charts/datadog-operator/templates/NOTES.txt#L44) that should prevent the deployment of unsupported configs (operator v1 with DatadogAgent v1alpha1 stored or old operator with the new v2alpha1).
We also documented the migration process if you are running the Datadog Operator `0.8.4` and the `v1alpha1` DatadogAgent here https://github.com/DataDog/helm-charts/tree/main/charts/datadog-operator#migration
Using a conversion webhook (with cert manager) this should allow you to migrate your setup.

Please let me know if you have any question or run into issues, we are trying to gather feedback on the migration and onboarding process.

**Comment by CharlyF** [comment]
I will be closing this as it is fixed. Please let us know if you run into any more issues.
